src/sql/src/plan/query.rs

```diff
@@ -1428,31 +1428,71 @@ pub fn plan_ctes(
 
             // Plan all CTEs and validate the proposed types.
             for cte in ctes.iter() {
-                let cte_name = normalize::ident(cte.name.clone());
                 let (val, _scope) = plan_nested_query(qcx, &cte.query)?;
+
                 // Validate that the derived and proposed types are the same.
-                let typ = qcx.relation_type(&val);
-                // TODO: Use implicit casts to convert among types rather than error.
-                if !typ.subtypes(qcx.ctes[&cte.id].desc.typ()) {
-                    let declared_typ = qcx.ctes[&cte.id]
-                        .desc
-                        .typ()
+                let derived_typ = qcx.relation_type(&val);
+                let proposed_typ = qcx.ctes[&cte.id].desc.typ();
+
+                let err = |proposed_typ: &RelationType, derived_typ: RelationType| {
+                    let cte_name = normalize::ident(cte.name.clone());
+                    let proposed_typ = proposed_typ
                         .column_types
                         .iter()
                         .map(|ty| qcx.humanize_scalar_type(&ty.scalar_type))
                         .collect::<Vec<_>>();
-                    let inferred_typ = typ
+                    let inferred_typ = derived_typ
                         .column_types
                         .iter()
                         .map(|ty| qcx.humanize_scalar_type(&ty.scalar_type))
                         .collect::<Vec<_>>();
                     Err(PlanError::RecursiveTypeMismatch(
                         cte_name,
-                        declared_typ,
+                        proposed_typ,
                         inferred_typ,
-                    ))?;
+                    ))
+                };
+
+                let all_keys = proposed_typ.keys.iter().all(|key1| {
+                    derived_typ
+                        .keys
+                        .iter()
+                        .any(|key2| key1.iter().all(|k| key2.contains(k)))
+                });
+
+                if !all_keys {
+                    return err(proposed_typ, derived_typ);
```

Comment:
I think this was pre-existing, but if these don't match, the error message will talk only about the scalar types when in fact it is a key mismatch, no?

Reply:
This was actually impossible to trigger because WMR CTEs don't support keys. Converted the existence of any key into an error so that if we support it, we'll have to fix this error (assuming we write a test 😄)